iree/turbine/kernel/wave/scheduling/verifier.py

```diff
@@ -0,0 +1,432 @@
+        return (
+            node_rrt_val.shape[0]
+            if node_rrt_val is not None and node_rrt_val.size > 0
+            else 0
+        )
+
+    def _apply_node_operation(
+        self,
+        node: fx.Node,
+        start_cycle: int,
+        node_rrt_getter: NodeRRTGetter,
+        operation: Callable[[np.ndarray, np.ndarray], np.ndarray],
+    ) -> None:
+        node_rrt_val = node_rrt_getter(node)
+        if node_rrt_val is None or node_rrt_val.size == 0:
+            return
+        for i in range(node_rrt_val.shape[0]):
+            cycle = (start_cycle + i) % self.T
+            self.RT_global[cycle, :] = operation(
+                self.RT_global[cycle, :], node_rrt_val[i, :]
+            )
+
+    def add_node(
+        self, node: fx.Node, start_cycle: int, node_rrt_getter: NodeRRTGetter
+    ) -> None:
+        self._apply_node_operation(
+            node, start_cycle, node_rrt_getter, lambda x, y: x + y
+        )
+
+    def remove_node(
+        self, node: fx.Node, start_cycle: int, node_rrt_getter: NodeRRTGetter
+    ) -> None:
+        self._apply_node_operation(
+            node, start_cycle, node_rrt_getter, lambda x, y: x - y
+        )
+
+    def can_add_node(
+        self, node: fx.Node, start_cycle: int, node_rrt_getter: NodeRRTGetter
+    ) -> bool:
+        node_rrt_val = node_rrt_getter(node)
+        if node_rrt_val is None or node_rrt_val.size == 0:
+            return True
+        for i in range(node_rrt_val.shape[0]):
+            cycle = (start_cycle + i) % self.T
+            if np.any(
+                self.RT_global[cycle, :] + node_rrt_val[i, :] > self.resource_limits
+            ):
+                return False
+        return True
+
+
+class ScheduleDependencyGraph:
+    """Represents and manages the dependency relationships between scheduled operations."""
+
+    def __init__(self, nodes: List[fx.Node], edges: RawEdgesList = None):
+        self.nodes = list(nodes)
+        self.edges = edges
+        self._adj = self._build_adjacency_list(edges, is_successors=True)
+        self._pred_adj = self._build_adjacency_list(edges, is_successors=False)
+
+    def _build_adjacency_list(
+        self, edges_input: RawEdgesList, is_successors: bool
+    ) -> Dict[fx.Node, List[fx.Node]]:
+        adj_map = {node: [] for node in self.nodes}
+        for edge in edges_input:
+            if is_successors:
+                if edge._from in adj_map and edge._to not in adj_map[edge._from]:
+                    adj_map[edge._from].append(edge._to)
+            else:
+                if edge._to in adj_map and edge._from not in adj_map[edge._to]:
+                    adj_map[edge._to].append(edge._from)
+        return adj_map
+
+    def get_successors(self, node: fx.Node) -> List[fx.Node]:
+        return self._adj.get(node, [])
+
+    def get_predecessors(self, node: fx.Node) -> List[fx.Node]:
+        return self._pred_adj.get(node, [])
+
+    def has_edge(self, pred: fx.Node, succ: fx.Node) -> bool:
```

Comment:
I think, here we are doing linear search every time. 
Will it be useful in the future PR to convert it into a set, if edges are of higher order?

Reply:
Yes, good point this can be converted into a set to make it faster. I will change it.